pkg/api/admin/api/types.go

```diff
@@ -0,0 +1,215 @@
+package api
+
+const (
+	// APIVersion is the version of this API
+	APIVersion = "admin"
+)
+
+// OpenShiftManagedCluster complies with the ARM model of resource definition in
+// a JSON template.
+type OpenShiftManagedCluster struct {
+	Plan       *ResourcePurchasePlan `json:"plan,omitempty"`
+	Properties *Properties           `json:"properties,omitempty"`
+	ID         *string               `json:"id,omitempty"`
+	Name       *string               `json:"name,omitempty"`
+	Type       *string               `json:"type,omitempty"`
+	Location   *string               `json:"location,omitempty"`
+	Tags       map[string]*string    `json:"tags"`
+
+	Config *Config `json:"config,omitempty"`
+}
+
+// ResourcePurchasePlan defines the resource plan as required by ARM for billing
+// purposes.
+type ResourcePurchasePlan struct {
+	Name          *string `json:"name,omitempty"`
+	Product       *string `json:"product,omitempty"`
+	PromotionCode *string `json:"promotionCode,omitempty"`
+	Publisher     *string `json:"publisher,omitempty"`
+}
+
+// Properties represents the cluster definition.
+type Properties struct {
+	// ProvisioningState (out): current state of the OSA resource.
+	ProvisioningState *ProvisioningState `json:"provisioningState,omitempty"`
+
+	// OpenShiftVersion (in): OpenShift version to be created/updated, e.g.
+	// `v3.11`.
+	OpenShiftVersion *string `json:"openShiftVersion,omitempty"`
+
+	// PublicHostname (in,optional): Optional user-specified FQDN for OpenShift
+	// API server.  If specified, after OSA cluster creation, user must create a
+	// PublicHostname CNAME record forwarding to the returned FQDN value.
+	PublicHostname *string `json:"publicHostname,omitempty"`
+
+	// FQDN (in): FQDN for OpenShift API server.  User-specified FQDN for
+	// OpenShift API server loadbalancer internal hostname.
+	FQDN *string `json:"fqdn,omitempty"`
+
+	// NetworkProfile (in): Configuration for OpenShift networking.
+	NetworkProfile *NetworkProfile `json:"networkProfile,omitempty"`
+
+	// RouterProfiles (in,optional/out): Configuration for OpenShift router(s).
+	RouterProfiles []RouterProfile `json:"routerProfiles,omitempty"`
+
+	// AgentPoolProfiles (in): configuration of OpenShift cluster VMs.
+	AgentPoolProfiles []AgentPoolProfile `json:"agentPoolProfiles,omitempty"`
+
+	// AuthProfile (in): configures OpenShift authentication
+	AuthProfile *AuthProfile `json:"authProfile,omitempty"`
```

Comment:
@jim-minter while on removing Secret from AADIdentityProvider, I am still trying to figure out whether exposing most of these options (parts of Properties) makes any sense for an SRE to manipulate.

Reply:
I think in the future customer will be able to change these via customerAPI. Question is do we want to have this now here or not.